nutils/evaluable.py

```diff
@@ -327,50 +328,75 @@ def _deep_flatten_constants(self):
-            if replacements:
-                self = util.shallow_replace(lambda key: replacements.get(key) if isinstance(key, LoopConcatenate) else None)(self)
+    def _combine_loops(self, candidates=None):
+        if candidates is None:
+            candidates = self._loop_deps
+        candidates = candidates.copy()
+
+        while candidates:
+            # Select the loops from candidates that can be combined in variable
+            # `loops`. We start with a subset of candidates, `loops_sans_deps`,
+            # with loops that are not a dependency of another loop in
+            # `candidates`. This ensures that the `candidates` do not disappear
+            # from `self` other than the deliberate combining performed here.
+            # To illustrate this, consider the following abstract loop
+            # structure
+            #
+            #     A: Add
+            #       B: LoopSum, loop_index=i
+            #         C: Add
+            #           D: LoopSum, loop_index=j
+            #             ..., does not depend on i
+            #           E: LoopSum, loop_index=j
+            #             ..., does not depend on i
+            #       F: LoopSum, loop_index=i
+            #         ...
+            #
+            # Loops D and E are invariant loops of B. The default candidates of
+            # A are B, D, E and F. By combing D and E first, B would be replaced by B',
+            #
+            #     A: Add
+            #       B': LoopSum, loop_index=i
+            #         C': Add
+            #           D': ArrayFromTuple, index=0
+            #             DE': LoopTuple(D, E), loop_index=j
+            #           E': ArrayFromTuple, index=0
+            #             DE'
+            #       F: LoopSum, loop_index=i
+            #         ...
+            #
+            # which is not in the set of candidates and we'd miss the
+            # opportunity to combine B' with F. Or worse: we combine B with F,
+            # ignore the former (`ArrayFromTuple`) *and* keep B'.
+            loops_sans_deps = candidates.copy()
+            for loop in candidates:
+                loops_sans_deps.difference_update(loop._loop_deps - util.IDSet([loop]))
+            index = next(iter(loops_sans_deps)).index
+            loops = util.IDSet([loop for loop in loops_sans_deps if loop.index == index])
+            candidates.difference_update(loops)
+            assert loops
+
+            if len(loops) >= 2:
+                replacements = util.IDDict()
+                combined = _LoopTuple(tuple(loops), index.name, index.length)
+                combined = combined._combine_loops(combined.body_arg._loop_deps - combined._loop_deps)
+                for i, loop in enumerate(loops):
+                    kwargs = {}
+                    replacements[loop] = ArrayFromTuple(combined, i, loop.shape, loop.dtype)
+                self = util.shallow_replace(replacements.get, self)
```

Comment:
Can't we wait till the end to apply all replacements at once?

Reply:
No, I don't think so. `shallow_replace` stops at the shallowest replacement, so we'd miss replacements that are a dependency of the shallowest replacement.